activation/poet.go

```diff
@@ -151,27 +156,52 @@ func (c *HTTPPoetClient) PowParams(ctx context.Context) (*PoetPowParams, error)
 	}, nil
 }
 
+func (c *HTTPPoetClient) CertifierInfo(ctx context.Context) (*url.URL, []byte, error) {
+	info, err := c.info(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	certifierInfo := info.GetCertifier()
+	if certifierInfo == nil {
+		return nil, nil, errors.New("poet doesn't support certifier")
+	}
+	url, err := url.Parse(certifierInfo.Url)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parsing certifier address: %w", err)
+	}
+	return url, certifierInfo.Pubkey, nil
+}
+
 // Submit registers a challenge in the proving service current open round.
 func (c *HTTPPoetClient) Submit(
 	ctx context.Context,
 	deadline time.Time,
 	prefix, challenge []byte,
 	signature types.EdSignature,
 	nodeID types.NodeID,
-	pow PoetPoW,
+	auth PoetAuth,
 ) (*types.PoetRound, error) {
 	request := rpcapi.SubmitRequest{
 		Prefix:    prefix,
 		Challenge: challenge,
 		Signature: signature.Bytes(),
 		Pubkey:    nodeID.Bytes(),
-		Nonce:     pow.Nonce,
-		PowParams: &rpcapi.PowParams{
-			Challenge:  pow.Params.Challenge,
-			Difficulty: uint32(pow.Params.Difficulty),
-		},
-		Deadline: timestamppb.New(deadline),
+		Deadline:  timestamppb.New(deadline),
+	}
+	if auth.PoetPoW != nil {
+		request.PowParams = &rpcapi.PowParams{
+			Challenge:  auth.PoetPoW.Params.Challenge,
+			Difficulty: uint32(auth.PoetPoW.Params.Difficulty),
+		}
+		request.Nonce = auth.PoetPoW.Nonce
+	}
+	if auth.PoetCert != nil {
+		request.Certificate = &rpcapi.SubmitRequest_Certificate{
+			Data:      auth.PoetCert.Data,
+			Signature: auth.PoetCert.Signature,
+		}
 	}
```

Comment:
What if both or neither are `nil`?

Reply:
Neither should happen because PoW is always and only ran if the cert is unavailable. However, assuming the caller is bad and it happens:
- both `nil` - the poet will reject registration,
- both are set - the poet will use the certificate.

Do you know a pattern/construct in Go to express a variant to enforce that always only one of them is set?

Reply:
> Do you know a pattern/construct in Go to express a variant to enforce that always only one of them is set?

Using 2 methods instead of one 🙂